File: 826-Most Profit Assigning Work.py
class Solution(object):
    def maxProfitAssignment(self, difficulty, profit, worker):
        """
        :type difficulty: List[int]
        :type profit: List[int]
        :type worker: List[int]
        :rtype: int
        """
        arr = sorted(zip(difficulty, profit))
        
        worker.sort()
        idx = 0
        maxP = 0
        res = 0
        
        for cap in worker:
            
            while idx < len(arr) and arr[idx][0] <= cap:
                maxP = max(maxP, arr[idx][1])
                idx += 1
            res += maxP
        
        return res

File: C++/826-Most Profit Assigning Work.cpp
class Solution {
public:
    int maxProfitAssignment(vector<int>& difficulty, vector<int>& profit, vector<int>& worker) {
        vector<pair<int, int>> arr;
        int N = difficulty.size();
        map<int, int> mp;

        for (int i=0; i<N; i++) arr.push_back({difficulty[i], profit[i]});

        sort(arr.begin(), arr.end());

        int maxVal = 0;
        for (int i=0; i<N; i++) {
            maxVal = max(maxVal, arr[i].second);
            mp[arr[i].first] = maxVal;
        }

        int res = 0, currMax = 0;
        for (int i=0; i<worker.size(); i++) {
            auto it = mp.upper_bound(worker[i]);
            if (it != mp.begin()) {
                it--;
                res += it->second;
            }
        }
        return res;
    }
};